Introduce bulk loading to StorageEmbeddedRocksDB

canhld94 opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):

- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

- Introduce bulk loading to StorageEmbeddedRocksDB by creating and ingesting SST file instead of relying on rocksdb build-in memtable. This help to increase importing speed, especially for long-running insert query to StorageEmbeddedRocksDB tables. Also, introduce `StorageEmbeddedRocksDB` table settings.

More detail in #59128, close #59128, close #58102

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

canhld94 committed:
Introduce EmbeddedRocksDBBulkSink

No mem-table: an SST file is built from chunk, then import to rocksdb.

Also add RocksDBSettings as table level settings for StorageEmbeddedRocksDB

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #59163 from canhld94/ClickHouse:rocksdb_bypass_memtable -->
*This is an automated comment for commit fb17d0d16939434be0990b18cfdd1fecc6227bb1 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59163/fb17d0d16939434be0990b18cfdd1fecc6227bb1/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

vdimir committed:
Add optimize_for_bulk_insert to embedded-rocksdb.md

vdimir commented:
Added entry to markdown file in docs/ to run a CI

canhld94 committed:
clean up and add some comments

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 commented:
@vdimir thank you! We still need can-be-tested label.

canhld94 committed:
fix document spell check

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
no rocksdb in fast test

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
fix tests

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
Merge branch 'master' into rocksdb_bypass_memtable

Try fixing weird CI errors

canhld94 committed:
hotfix rocksdb bulk sink fails with user defined query id

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
hotfix: handling empty chunk

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
rocksdb: enforce batch size for bulk insert

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
rocksdb: implement ALTER SETTING

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
fix typo

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
better rocksdb test

- add test for alter modify setting
- make sure bulk insert kick in

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

vdimir commented on src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp:
```diff
@@ -0,0 +1,235 @@
+    /// to have a distinct path. Also we cannot use query id as directory name here, because it could be defined by user and not suitable
+    /// for directory name
+    auto base_directory_name = sipHash128String(getContext()->getCurrentQueryId());
+    insert_directory_queue = fs::path(storage.getDataPaths()[0]) / (base_directory_name + "-" + getRandomASCIIString(8));
```
Have you considered storing temporary files in [`tmp_path`](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#tmp-path)? As I see when we store it directly in storage path we can avoid copying. But storing in tmp also has some advantages, for example it can be cleaned up on startup in case of server failures.

canhld94 replied:
I still prefer to write temporary files in same disk. But I agree that we need to handle garbage directories when server failures. I think we can do this: temporary insert directory will have a prefix (for example, `tmp_bulk_sink_`), then in `StorageEmbeddedRocksDB::initDB()` we can scan temporary insert directory and remove them before initializing rocksdb database. @vdimir Wdyt about this approach?

canhld94 replied:
I still prefer to have temporary files in same disk. But I agree we need to handle garbage when server failures. I think we can do it by having a specific prefix for temporary insert directory and then clean it in `StorageEmbeddedRocksDB::initDB()`. Wdyt?

vdimir commented on src/Storages/RocksDB/EmbeddedRocksDBBulkSink.h:
```diff
@@ -0,0 +1,67 @@
+    Serializations serializations;
+
+    /// For squashing chunks
+    std::vector<Chunk> chunks;
```
Shouldn't it be protected from concurrent access?

canhld94 replied:
I don't see any possible concurrent access here. It's not accessible anywhere outside of `EmbeddedRocksDBBulkSink`. Can you give an example?

canhld94 replied:
I don't see any possible concurrent access. `chunks` is not accessible outside of `EmbeddedRocksDBBulkSink`. Can you give an example?

vdimir commented on tests/queries/0_stateless/02956_rocksdb_bulk_sink.sql:
```diff
@@ -0,0 +1,6 @@
+-- Tags: no-ordinary-database, no-fasttest
```
Effectevely it's the same, but would work if we enabled rocksdb in fast tests and do not require a comment with the reason

```suggestion
-- Tags: no-ordinary-database, use-rocksdb
```

vdimir commented on src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp:
```diff
@@ -0,0 +1,235 @@
+        return status;
+
+    auto rows = perm.size();
+    WriteBufferFromOwnString wb_value;
```
unused 
```suggestion
```

vdimir commented on tests/queries/0_stateless/02956_rocksdb_bulk_sink.sql:
```diff
@@ -0,0 +1,6 @@
+-- Tags: no-ordinary-database, no-fasttest
+-- Tag no-ordinary-database: Sometimes cannot lock file most likely due to concurrent or adjacent tests, but we don't care how it works in Ordinary database
+-- Tag no-fasttest: In fasttest, ENABLE_LIBRARIES=0, so rocksdb engine is not enabled by default
+CREATE TABLE rocksdb_worm (key UInt64, value UInt64) ENGINE = EmbeddedRocksDB() PRIMARY KEY key SETTINGS optimize_for_bulk_insert = 1;
+INSERT INTO rocksdb_worm SELECT number, number+1 FROM numbers_mt(1000) SETTINGS max_insert_threads = 2;
```
Also consider making test a bit more sophisticated, for example testing several concurrent INSERTS (via `.sh` test`)

canhld94 replied:
Ok, I will write a more comprehensive test.

canhld94 replied:
Yes, I will make more comprehensive test.

canhld94 replied:
Added more tests

vdimir reviewed (commented):
Thank you for your PR! Looks good overall, the main questions are about using storage directory for temporary file, squashing inside the `Sink`, and maybe lack of different scenarios in test, see comments, please

canhld94 commented on src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp:
```diff
@@ -0,0 +1,235 @@
+
+        status = sst_file_writer.Put(keys.getDataAt(row).toView(), values.getDataAt(row).toView());
+
+        /// There could be duplicated keys in chunk, thus Put may give IsInvalidArgument. This is ok, as we're certain that
```
It's about performance. `Put` will compare keys before insertion, so I think we don't need to do it before `Put`. Regarding concern about `IsInvalidArgument`, can we compared the error message and "Keys must be added in strict ascending order.". Wdyt?
```cpp
    if (file_info.num_entries == 0) {
      file_info.smallest_key.assign(user_key.data(), user_key.size());
    } else {
      if (internal_comparator.user_comparator()->Compare(
              user_key, file_info.largest_key) <= 0) {
        // Make sure that keys are added in order
        return Status::InvalidArgument(
            "Keys must be added in strict ascending order.");
      }
    }
```

canhld94 replied:
It's about performance. `Put` already compares keys before insertion, so I think we don't need to do it here.

```cpp
    if (file_info.num_entries == 0) {
      file_info.smallest_key.assign(user_key.data(), user_key.size());
    } else {
      if (internal_comparator.user_comparator()->Compare(
              user_key, file_info.largest_key) <= 0) {
        // Make sure that keys are added in order
        return Status::InvalidArgument(
            "Keys must be added in strict ascending order.");
      }
```

Regarding concern about `IsInvalidArgument`, I think if status code is `IsInvalidArgument` then we also compare the status message to  "Keys must be added in strict ascending order." to verify it's indeed indicate a duplicated key. Do you think this approach is good?

canhld94 replied:
I changed the code to deduplicate keys before `Put`. The previous implementation is not correct because we suppose to select the last value of the if key is duplicated.

canhld94 commented on src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp:
```diff
@@ -0,0 +1,235 @@
+    auto [serialized_key_column, serialized_value_column] = serializeChunks(to_written);
+    auto path = getTemporarySSTFilePath();
+    if (auto status = buildSSTFile(path, *serialized_key_column, *serialized_value_column); !status.ok())
+        throw Exception(ErrorCodes::ROCKSDB_ERROR, "RocksDB write error: {}", status.ToString());
```
In the destructor of `EmbeddedRocksDBBulkSink` we will remove the working directory, so I think don't need to remove `path` here?

canhld94 replied:
In the destructor of `EmbeddedRocksDBBulkSink` we will remove the temporary insert directory anyway, so I think we don't need to explicitly remove `path` here?

canhld94 commented on src/Storages/RocksDB/EmbeddedRocksDBBulkSink.h:
```diff
@@ -0,0 +1,67 @@
+    String getTemporarySSTFilePath();
+
+    /// Squash chunks to a minimum size
+    std::vector<Chunk> squash(Chunk chunk);
```
The squashing logic here is different from `SquashingChunksTransform` logic. We don't create a new big chunk but just group some chunks together into a vector, which is much cheaper. `serializeChunks` will build final key-value columns from those chunks.

canhld94 replied:
Squashing logic here is different from `SquashingChunksTransform`. We don't create a new big chunk but grouping chunks in to a vector until we have enough rows, which is much cheaper. We can do this because `serializeChunks` can create key-value columns from multiple chunks.

canhld94 commented on tests/queries/0_stateless/02956_rocksdb_bulk_sink.sql:
```diff
@@ -0,0 +1,6 @@
+-- Tags: no-ordinary-database, no-fasttest
+-- Tag no-ordinary-database: Sometimes cannot lock file most likely due to concurrent or adjacent tests, but we don't care how it works in Ordinary database
```
I just copy this line from other rocksdb tests.. May be it's outdated now...

canhld94 replied:
I copy this line from other rocksdb tests...  May be it's already outdated.

canhld94 reviewed (commented):
@vdimir thanks for your review.

canhld94 committed:
Apply suggestions from code review

Co-authored-by: vdimir <vdimir@clickhouse.com>

canhld94 committed:
better way to deduplicate keys while creating sst files

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
Merge branch 'master' into rocksdb_bypass_memtable

Resolve conflicts

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
fix clang-tidy and better test

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
update documents

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
Fix style

canhld94 commented:
@vdimir sorry for delayed for a long time...
We've been running this PR in our prod and discover some bugs. Till now it's all good and stable.
I've addressed your concerns as well as adding more test for it.
Could you please look at this again? Thank you!

canhld94 commented on src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp:
```diff
@@ -0,0 +1,243 @@
+            {
+                for (size_t idx = 0; idx < columns.size(); ++idx)
+                    serializations[idx]->serializeBinary(*columns[idx], i, idx == primary_key_pos ? writer_key : writer_value, {});
+                writeChar('\0', writer_key);
+                writeChar('\0', writer_value);
```
@vdimir the writer buffer in two places are being used differently:

- In `EmbeddedRocksDBSink`, it is used to serialised one key-value pair then to call `WriteBatch`.
- In `EmbeddedRocksDBBulkSink`, it is used to serialised keys and values to two `ColumnString`. That's why after serialising I added trailing zeros, because row in `ColumnString` should be null-terminated.

I wil check more but it's quite certain that the two serialisation logic are same, because in `ColumnString` the last null character is not part of the string. I will add a test for it as well (e.g. first import some key-value pairs not using bulk-sink, then enable bulk-sink and import pairs with the key but different value, and make sure that the later values override the prior).

> Actually, it might be good to unify the serialization code between EmbeddedRocksDBSink and EmbeddedRocksDBBulkSink so that it's clear they are serialized in the same way.

I think we can change the `EmbeddedRocksDBSink` logic, not serialise -> write batch -> serialise -> ... but serialise all records first, then write batch. RocksDB write batch has API to write multiple key-value pairs at once. Wdyt?

vdimir replied:
> I think we can change the EmbeddedRocksDBSink logic, not serialise -> write batch -> serialise -> ... but serialise all records first, then write batch. RocksDB write batch has API to write multiple key-value pairs at once. Wdyt?

Wouldn't it increase memory consuption since we will have to have whole batch serialized to string in memory instead of single row in `WriteBufferWithOwnString`

canhld94 replied:
Yes, that's also my concern. But it's the only approach if we want to unify the serialisation logic in those two class.

May be as you said, we can left it for future improvement. I will just add a test for now.

canhld94 replied:
Added test to compare bulk sink vs normal sink serialisation.

canhld94 commented on src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp:
```diff
@@ -0,0 +1,243 @@
+    return {std::move(serialized_key_column), std::move(serialized_value_column)};
+}
+
+void EmbeddedRocksDBBulkSink::consume(Chunk chunk_)
+{
+    std::vector<Chunk> to_written = squash(std::move(chunk_));
+
+    if (to_written.empty())
+        return;
```
I don't think it can be called from multiple threads. If we have multiple threads for insert, it will create multiple `EmbeddedRocksDBBulkSink` objects. I will check more.

vdimir replied:
Why `file_counter` is `std::atomic_size_t`, not `size_t` then?

canhld94 replied:
It should be a mistake, I will change to `size_t`.

canhld94 replied:
changed to `size_t`

canhld94 replied:
I recalled why it's `std::atomic_size` then. The inserting flow including two major steps: (1) create serialised key-value chunk and sort the chunk by key, and (2) create SST file and import it. (1) is compute-heavy, while (2) is I/O heavy, so I had an idea to overlap them. However, later I think that may be it's no different from having `max_insert_threads = 2`, so it's on hold now.

canhld94 committed:
Update src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp

Co-authored-by: vdimir <vdimir@clickhouse.com>

canhld94 committed:
address reviews

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 commented on src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp:
```diff
@@ -0,0 +1,243 @@
+    /// to have a distinct path. Also we cannot use query id as directory name here, because it could be defined by user and not suitable
+    /// for directory name
+    auto base_directory_name = sipHash128String(getContext()->getCurrentQueryId());
+    insert_directory_queue = fs::path(storage.getDataPaths()[0]) / (base_directory_name + "-" + getRandomASCIIString(8));
```
Addd `tmp_insert_` prefix

canhld94 committed:
update comments

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
Merge branch 'master' into rocksdb_bypass_memtable

Try fix CI

canhld94 commented:
Failed stateless test:

```
02992_all_columns_should_have_comment
02999_scalar_subqueries_bug_2
```

Non of them looks relates to the this PR.

vdimir reviewed (approved):
Thanks for taking the time to work on the PR!

alexey-milovidov commented:
Sorry, one very small problem has been found:

TSan report on this line:

```
ingest_options.move_files = true;
```

Please resubmit.

alexey-milovidov commented:
https://github.com/ClickHouse/ClickHouse/pull/63316#issuecomment-2092033228
